lib/commands/toggleList.js

```diff
@@ -0,0 +1,103 @@
+// @flow
+import { Editor, Node, NodeEntry, Path, Range } from 'slate';
+import type { Options } from '..';
+import { unwrapList, wrapInList } from '.';
+import { isList, isItem, getItemsAtRange } from '../utils';
+
+export const isListOrItem = (options: Options) => (node: Node): boolean =>
+    isList(options)(node) || isItem(options)(node);
+
+const isSameLevel = (nodeEntries: Array<NodeEntry>): boolean => {
+    if (nodeEntries.length === 0) {
+        return true;
+    }
+
+    const referenceDepth = nodeEntries[0][1].length;
+
+    return !nodeEntries.some(
+        ([, nodeEntryPath]) => nodeEntryPath.length !== referenceDepth
+    );
+};
+
+const unwrapAllLists = (options: Options) => (
+    editor: Editor,
+    nodes: Array<NodeEntry>
+) => {
+    const ancestorDescendantsLists = nodes
+        .filter(([node]) => isList(options)(node))
+        .reverse();
+
+    const descendantListsPathRefs = ancestorDescendantsLists.map(
+        ([, descendantPath]) => Editor.pathRef(editor, descendantPath)
+    );
+
+    // Unwrap all nested lists
+    Editor.withoutNormalizing(editor, () => {
+        descendantListsPathRefs.forEach(descendantListPathRef => {
+            unwrapList(options)(editor, descendantListPathRef.current);
+        });
+    });
+
+    // Unwrap common ancestor
+    unwrapList(options)(editor);
+};
+
+/**
+ * Toggle list on the selected range.
+ */
+export const toggleList = (options: Options) => (editor: Editor): void => {
+    const range = editor.selection;
+    const [startElement, startElementPath] = Editor.parent(
+        editor,
+        Range.start(range)
+    );
+    const [endElement, endElementPath] = Editor.parent(
+        editor,
+        Range.end(range)
+    );
+
+    // -------- SINGLE BLOCK ---------------------------------------------------
+    // The selection is in a single block.
+    // Let's unwrap just the block, not the whole list.
+    if (startElement === endElement) {
+        if (getItemsAtRange(options)(editor).length > 0) {
+            unwrapList(options)(editor);
+        } else {
+            wrapInList(options)(editor);
+        }
+        return;
+    }
+
+    // -------- NOT A SINGLE BLOCK -------------------------------------------
+    const ancestorPath = Path.common(startElementPath, endElementPath);
+    const ancestor = Node.get(editor, ancestorPath);
+
+    const ancestorDescendants = Array.from(
+        Editor.nodes(editor, {
+            at: editor.selection,
+            match: isListOrItem(options),
+        })
+    ).filter(([, descendantPath]) => descendantPath >= ancestorPath);
+    const ancestorDescendantItems = ancestorDescendants.filter(
+        ([itemOrListNode]) => isItem(options)(itemOrListNode)
+    );
```

Comment:
You don't need to necessarily do this but try to check `for .. of` instead of `Array.from(`